src/lib.rs

```diff
@@ -284,26 +287,32 @@ impl<R: Read + Seek> TwoBitFile<R> {
     ) -> Result<String> {
         const NUC: &[u8; 4] = b"TCAG";
 
-        let seq = self.sequences.query(chr)?;
+        let seq = self.sequences.query(&chr)?;
         let reader = &mut self.reader;
 
-        let start = match range.start_bound() {
-            Bound::Included(&v) => v,
-            Bound::Excluded(&v) => v + 1,
-            Bound::Unbounded => 0,
+        let start = {
+            let pos = match range.start_bound() {
+                Bound::Included(&v) => v,
+                Bound::Excluded(&v) => v + 1,
+                Bound::Unbounded => 0,
+            };
+            min(pos, seq.length)
         };
-        let end = match range.end_bound() {
-            Bound::Included(&v) => v + 1,
-            Bound::Excluded(&v) => v,
-            Bound::Unbounded => seq.length,
+        let end = {
+            let pos = match range.end_bound() {
+                Bound::Included(&v) => v + 1,
+                Bound::Excluded(&v) => v,
+                Bound::Unbounded => seq.length,
+            };
+            min(pos, seq.length)
         };
 
-        let first_byte = start / 4;
-        reader.seek(SeekFrom::Start(seq.offset))?; // beginning of the DNA sequence
-        reader.seek(SeekFrom::Current(first_byte as _))?; // position where we want to start reading
         if start >= end {
             return Ok(String::new()); // trivial case, empty return result
         }
+        let first_byte = start / 4;
+        reader.seek(SeekFrom::Start(seq.offset))?; // beginning of the DNA sequence
+        reader.seek(SeekFrom::Current(first_byte as _))?; // position where we want to start reading
 
         let length = end - start;
         let mut out = Vec::with_capacity(length);
```

Comment:
This is old code, but the github clippy integration test fails now:

```
calling `set_len()` immediately after reserving a buffer creates uninitialized values
initialize the buffer or wrap the content in `MaybeUninit`
for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#uninit_vec
```

I get no complaints from clippy when I check this locally.

Reply:
I just switched to nightly. Clippy complains now.

Reply:
I did some more experiments with the generated assembly code for `vec![0_u8; length]`. Without optimizations this translates into a `memset` call. But with optimizations it translates into  `callq *__rust_alloc_zeroed@GOTPCREL(%rip)`. So this should be even faster than regular `memset`. I changed the code to zero-initialize the buffer to make clippy happy.

Reply:
I think you could simply extend or push to the `Vec`. The way you do it is just using a local variable to keep track of the length.